backends/dpdk/dpdkAsmOpt.cpp

```diff
@@ -163,4 +163,32 @@ const IR::IndexedVector<IR::DpdkAsmStatement> *RemoveLabelAfterLabel::removeLabe
     }
     return new_l;
 }
+
+const IR::Node* RemoveUnusedMetadataFields::preorder(IR::DpdkAsmProgram *p) {
+    auto usedFieldsCollector = new CollectUsedMetadataField();
+    usedFieldsCollector->setCalledBy(this);
+    p->apply(*usedFieldsCollector);
+    auto used_fields = usedFieldsCollector->getUsedFields();
+
+    IR::IndexedVector<IR::DpdkStructType> usedStruct;
+    for (auto st : p->structType) {
+        IR::IndexedVector<IR::StructField> usedMetadataFields;
+        for (auto field : st->fields) {
```

Comment:
We should use some early exit condition to avoid looping on header structs, Probably, annotations can be checked. The annotation "__metadata__" is added to the metadata struct and since there is a single metadata struct at this point, this loop will be taken only once.

Reply:
done